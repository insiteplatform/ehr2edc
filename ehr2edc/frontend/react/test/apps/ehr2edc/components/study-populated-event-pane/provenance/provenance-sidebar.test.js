import React from "react"
import {
    mapDispatchToProps,
    ProvenanceSidebar
} from "../../../../../../src/apps/ehr2edc/components/study-populated-event-pane/provenance/provenance-sidebar";
import {
    closeItemDetails,
    FETCH_ITEM_PROVENANCE,
    FETCH_SUBMITTED_ITEM_PROVENANCE
} from "../../../../../../src/apps/ehr2edc/actions/item-provenance.actions";

describe("ProvenanceSidebar rendering", () => {
    function resetDocument() {
        document.body.innerHTML = '';
    }

    function createPortalRoot() {
        const portalRoot = global.document.createElement('div');
        portalRoot.setAttribute('id', 'study-details-content-pane');
        const body = global.document.querySelector('body');
        body.appendChild(portalRoot);
    }

    beforeEach(() => {
        resetDocument();
        createPortalRoot();
    });

    test("Should not render with incomplete props", () => {
        // Expect:The component not to be rendered, when I mount the component with invalid props
        expect(mount(<ProvenanceSidebar/>)).toEqual({});
        expect(mount(<ProvenanceSidebar {...validProps} name={""}/>)).toEqual({});
        expect(mount(<ProvenanceSidebar {...validProps} name={undefined}/>)).toEqual({});
        expect(mount(<ProvenanceSidebar {...validProps} value={""}/>)).toEqual({});
        expect(mount(<ProvenanceSidebar {...validProps} value={undefined}/>)).toEqual({});
        expect(mount(<ProvenanceSidebar {...validProps} provenance={"notAnObject"}/>)).toEqual({});
        expect(mount(<ProvenanceSidebar {...validProps} provenance={undefined}/>)).toEqual({});
    });

    test("Should create a portal to #study-details-content-pane", () => {
        // Given: A root to attach the Sidebar to
        const portalRoot = global.document.querySelector('#study-details-content-pane');

        // When: I mount the component
        const comp = mount(<ProvenanceSidebar {...validProps}/>);

        // Then: The sidebar is added to the DOM
        expect(comp.find("Portal").first().prop("container")).toEqual(portalRoot);
    });

    test("Should render a ItemProvenance-component inside a Sidebar", () => {
        // Given: A ProvenanceSidebar
        const app = mount(<ProvenanceSidebar {...validProps}/>);

        // Then: The component should match the snapshot
        expect(app).toMatchSnapshot();
        // And: A Sidebar should be rendered
        const sidebar = app.find("Sidebar");
        expect(sidebar.exists()).toBe(true);
        // And: The Sidebar should contain an ItemProvenance
        expect(sidebar.exists("ItemProvenance")).toBe(true);
    })
});

describe("Actions", () => {
    it("Should dispatch the fetchItemProvenance action when item is in review and item identifiers are set as props", () => {
        // Given: A fetchItemProvenance action
        const fetchItemProvenance = jest.fn();
        const actions = {
            fetchItemProvenance: fetchItemProvenance
        };

        // When: Rendering the component in review with mandatory identifiers
        const component = mount(<ProvenanceSidebar inReview={true}
                                                   actions={actions}/>);
        component.setProps(identifiers);

        // Then: The fetchItemProvenance action is dispatched
        expect(fetchItemProvenance).toHaveBeenCalledWith({...identifiers});
    });
    it("Should dispatch the fetchSubmittedItemProvenance action when item is submitted and item identifiers are set as props", () => {
        // Given: A fetchSubmittedItemProvenance action
        const fetchSubmittedItemProvenance = jest.fn();
        const actions = {
            fetchSubmittedItemProvenance: fetchSubmittedItemProvenance
        };

        // When: Rendering the component reviewed with mandatory identifiers
        const component = mount(<ProvenanceSidebar inReview={false}
                                                   actions={actions}/>);
        component.setProps(identifiers);

        // Then: The fetchSubmittedItemProvenance action is dispatched
        expect(fetchSubmittedItemProvenance).toHaveBeenCalledWith({...identifiers});
    });
    it("Should dispatch the closeItemDetails action when clicking the icon", () => {
        // Given: A close action
        const closeItemDetails = jest.fn();
        const actions = {
            closeItemDetails: closeItemDetails
        };

        // When: Clicking the close button
        const component = mount(<ProvenanceSidebar {...validProps} actions={actions}/>);
        component.find("a.close-icon").simulate("click");

        // Then: The close-action is dispatched
        expect(closeItemDetails).toHaveBeenCalled();
    });
    it("Should call the close handler and dispatch the closeItemDetails action when unmounting the component", () => {
        // Given: A close handler and action
        const onCloseSpy = jest.spyOn(ProvenanceSidebar.prototype, "onClose");
        const closeItemDetails = jest.fn();
        const actions = {
            closeItemDetails: closeItemDetails
        };
        // And: The component is mounted
        const component = mount(<ProvenanceSidebar {...validProps}
                                                   actions={actions}/>);

        // When: The component is unmounted
        component.unmount();

        // Then: The close handler and action is called
        expect(closeItemDetails).toHaveBeenCalled();
        expect(onCloseSpy).toHaveBeenCalled();
    });
});

describe("mapDispatchToProps", () => {
    it("Should map the fetchItemProvenance action", () => {
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).actions.fetchItemProvenance(identifiers);

        expect(dispatch).toBeCalledWith(expect.objectContaining({
            type: FETCH_ITEM_PROVENANCE
        }));
    });
    it("Should map the fetchSubmittedItemProvenance action", () => {
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).actions.fetchSubmittedItemProvenance(identifiers);

        expect(dispatch).toBeCalledWith(expect.objectContaining({
            type: FETCH_SUBMITTED_ITEM_PROVENANCE
        }));
    });
    it("Should map the closeItemDetails action", () => {
        const dispatch = jest.fn();

        mapDispatchToProps(dispatch).actions.closeItemDetails();

        expect(dispatch).toBeCalledWith(closeItemDetails());
    });
});

const identifiers = {
    studyId: "studyId",
    eventId: "eventId",
    subjectId: "subjectId",
    itemId: "itemId"
};

const validProps = {
    actions: {
        fetchItemProvenance: jest.fn(),
        fetchSubmittedItemProvenance: jest.fn(),
        openItemDetails: jest.fn(),
        closeItemDetails: jest.fn(),
        toggleNavigationSidebar: jest.fn()
    },
    name: "Lab Test or Examination Short Name",
    value: "VAL",
    provenance: {
        labConcept: {
            concept: {
                code: "INSULIN"
            },
            component: "Component for insulin",
            method: "Radioimmunoassay",
            fastingStatus: "FASTING",
            specimen: "Blood"
        },
        startDate: "2019-07-08T13:35:00",
        endDate: "2019-07-09T13:35:00",
        quantitativeResult: {
            lowerLimit: 75,
            upperLimit: 100,
            value: 85,
            unit: "mg/dL"
        },
        qualitativeResult: {
            parsedInterpretation: 1,
            originalInterpretation: "positive"
        },
        vendor: "Lab"
    }
};