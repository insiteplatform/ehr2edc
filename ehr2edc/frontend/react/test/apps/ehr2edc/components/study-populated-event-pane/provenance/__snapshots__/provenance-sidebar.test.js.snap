// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProvenanceSidebar rendering Should render a ItemProvenance-component inside a Sidebar 1`] = `
<ProvenanceSidebar
  actions={
    Object {
      "closeItemDetails": [MockFunction],
      "fetchItemProvenance": [MockFunction],
      "fetchSubmittedItemProvenance": [MockFunction],
      "openItemDetails": [MockFunction],
      "toggleNavigationSidebar": [MockFunction],
    }
  }
  name="Lab Test or Examination Short Name"
  provenance={
    Object {
      "endDate": "2019-07-09T13:35:00",
      "labConcept": Object {
        "component": "Component for insulin",
        "concept": Object {
          "code": "INSULIN",
        },
        "fastingStatus": "FASTING",
        "method": "Radioimmunoassay",
        "specimen": "Blood",
      },
      "qualitativeResult": Object {
        "originalInterpretation": "positive",
        "parsedInterpretation": 1,
      },
      "quantitativeResult": Object {
        "lowerLimit": 75,
        "unit": "mg/dL",
        "upperLimit": 100,
        "value": 85,
      },
      "startDate": "2019-07-08T13:35:00",
      "vendor": "Lab",
    }
  }
  value="VAL"
>
  <SidebarPortal>
    <Portal
      container={
        <div
          id="study-details-content-pane"
        >
          <div
            class="sidebar grid-block vertical small-4 border-left slideInLeft slideOutRight"
          >
            <div
              class="sidebar-right"
            >
              <a
                class="float-right close-icon"
              >
                <i
                  class="fa fa-close"
                />
              </a>
              <div
                class="grid-block shrink vertical"
              >
                <h4
                  class="facts-title"
                >
                  Source Data Details
                </h4>
                <dl
                  class="margin-none"
                >
                  <h5>
                    Name
                  </h5>
                  <dd>
                    Lab Test or Examination Short Name
                  </dd>
                  <h5>
                    Submitted Value
                  </h5>
                  <dd>
                    VAL
                  </dd>
                </dl>
                <h5>
                  Source Data
                </h5>
                <p>
                  Loading...
                </p>
              </div>
            </div>
          </div>
        </div>
      }
    >
      <WaitForContainer
        container={
          <div
            id="study-details-content-pane"
          >
            <div
              class="sidebar grid-block vertical small-4 border-left slideInLeft slideOutRight"
            >
              <div
                class="sidebar-right"
              >
                <a
                  class="float-right close-icon"
                >
                  <i
                    class="fa fa-close"
                  />
                </a>
                <div
                  class="grid-block shrink vertical"
                >
                  <h4
                    class="facts-title"
                  >
                    Source Data Details
                  </h4>
                  <dl
                    class="margin-none"
                  >
                    <h5>
                      Name
                    </h5>
                    <dd>
                      Lab Test or Examination Short Name
                    </dd>
                    <h5>
                      Submitted Value
                    </h5>
                    <dd>
                      VAL
                    </dd>
                  </dl>
                  <h5>
                    Source Data
                  </h5>
                  <p>
                    Loading...
                  </p>
                </div>
              </div>
            </div>
          </div>
        }
      >
        <Portal
          containerInfo={
            <div
              id="study-details-content-pane"
            >
              <div
                class="sidebar grid-block vertical small-4 border-left slideInLeft slideOutRight"
              >
                <div
                  class="sidebar-right"
                >
                  <a
                    class="float-right close-icon"
                  >
                    <i
                      class="fa fa-close"
                    />
                  </a>
                  <div
                    class="grid-block shrink vertical"
                  >
                    <h4
                      class="facts-title"
                    >
                      Source Data Details
                    </h4>
                    <dl
                      class="margin-none"
                    >
                      <h5>
                        Name
                      </h5>
                      <dd>
                        Lab Test or Examination Short Name
                      </dd>
                      <h5>
                        Submitted Value
                      </h5>
                      <dd>
                        VAL
                      </dd>
                    </dl>
                    <h5>
                      Source Data
                    </h5>
                    <p>
                      Loading...
                    </p>
                  </div>
                </div>
              </div>
            </div>
          }
        >
          <Sidebar
            additionalClasses={
              Array [
                "small-4",
                "border-left",
                "slideInLeft",
                "slideOutRight",
              ]
            }
          >
            <div
              className="sidebar grid-block vertical small-4 border-left slideInLeft slideOutRight"
            >
              <div
                className="sidebar-right"
              >
                <a
                  className="float-right close-icon"
                  onClick={[Function]}
                >
                  <i
                    className="fa fa-close"
                  />
                </a>
                <div
                  className="grid-block shrink vertical"
                >
                  <ItemProvenance
                    item={
                      Object {
                        "name": "Lab Test or Examination Short Name",
                        "provenance": Object {
                          "endDate": "2019-07-09T13:35:00",
                          "labConcept": Object {
                            "component": "Component for insulin",
                            "concept": Object {
                              "code": "INSULIN",
                            },
                            "fastingStatus": "FASTING",
                            "method": "Radioimmunoassay",
                            "specimen": "Blood",
                          },
                          "qualitativeResult": Object {
                            "originalInterpretation": "positive",
                            "parsedInterpretation": 1,
                          },
                          "quantitativeResult": Object {
                            "lowerLimit": 75,
                            "unit": "mg/dL",
                            "upperLimit": 100,
                            "value": 85,
                          },
                          "startDate": "2019-07-08T13:35:00",
                          "vendor": "Lab",
                        },
                        "value": "VAL",
                      }
                    }
                  >
                    <h4
                      className="facts-title"
                    >
                      Source Data Details
                    </h4>
                    <dl
                      className="margin-none"
                    >
                      <h5>
                        Name
                      </h5>
                      <dd>
                        Lab Test or Examination Short Name
                      </dd>
                      <h5>
                        Submitted Value
                      </h5>
                      <dd>
                        VAL
                      </dd>
                    </dl>
                    <h5>
                      Source Data
                    </h5>
                    <p>
                      Loading...
                    </p>
                  </ItemProvenance>
                </div>
              </div>
            </div>
          </Sidebar>
        </Portal>
      </WaitForContainer>
    </Portal>
  </SidebarPortal>
</ProvenanceSidebar>
`;
